c_src/membrane_portaudio_plugin/pa_devices.c

```diff
@@ -1,31 +1,52 @@
 #include "pa_devices.h"
 
-UNIFEX_TERM list(UnifexEnv *env) {
+UNIFEX_TERM list(UnifexEnv *env)
+{
   Pa_Initialize();
   int numDevices = Pa_GetDeviceCount();
-  if (numDevices < 0) {
+  device devices[numDevices];
+  if (numDevices < 0)
+  {
     printf("\nERROR: Pa_CountDevices returned 0x%x\n", numDevices);
-    return list_result(env);
-  } else if (numDevices == 0) {
-    printf("\nNo audio devices found\n");
-  } else {
-    printf("\nAvailable audio devices:\n\n");
+    return list_result(env, devices, numDevices);
+  }
+  else if (numDevices == 0)
+  {
+    return list_result(env, devices, numDevices);
+  }
+  else
+  {
     int default_input_id = Pa_GetDefaultInputDevice();
     int default_output_id = Pa_GetDefaultOutputDevice();
-    for (int i = 0; i < numDevices; i++) {
+    for (int i = 0; i < numDevices; i++)
+    {
       const PaDeviceInfo *device_info = Pa_GetDeviceInfo(i);
-      const char *default_str =
-          i == default_input_id
-              ? " (default input)"
-              : i == default_output_id ? " (default output)" : "";
 
-      printf("%s%s\r\n\tid: %d\r\n\tmax_input_channels: "
-             "%d\r\n\tmax_output_channels: %d\r\n\n",
-             device_info->name, default_str, i, device_info->maxInputChannels,
-             device_info->maxOutputChannels);
+      if (i == default_input_id)
+      {
+        devices[i].default_device = DEFAULT_DEVICE_INPUT;
+      }
+      else if (i == default_output_id)
+      {
+        devices[i].default_device = DEFAULT_DEVICE_OUTPUT;
+      }
+      else
+      {
+        devices[i].default_device = DEFAULT_DEVICE_FALSE;
+      }
+
+      devices[i].name = malloc(strlen(device_info->name) + 1);
+      if (device_info->name != NULL)
+      {
+        strcpy(devices[i].name, device_info->name);
+      }
```

Comment:
If you allocate anything on the heap, you should free it. In this case, after the call to `list_result`. However, unifex does malloc and memcpy when converting to an erlang term, so it seems redundant.